src/modelfree/training/victim_envs.py

```diff
@@ -94,8 +112,65 @@ def get_curry_obs(self, env_idx=None):
             return self._obs[env_idx]
 
 
+class MultiCurryVecEnv(CurryVecEnv):
```

Comment:
Okay. My impression had been that it would make CurryVecEnv pretty messy and functionality-overloaded, since it'd need to have all the functionality for switching between policies even if there's only one, but I can merge them if you think they'd be better that way.

Reply:
I'm still in favour of merging them. `MultiCurryVecEnv` is a superset of `CurryVecEnv`'s functionality, so having both is redundant. I agree `MultiCurryVecEnv` is more complicated than `CurryVecEnv`, but it still seems easier to have to read and understand just one of them than both.

The strongest argument for keeping them separate is that `MultiCurryVecEnv` is only used in a fairly niche use case. But I think we still need that use case to be correct, so if we're worried about `MultiCurryVecEnv` we should just add more tests.

Reply:
Slight complication: I hadn't realized `TransparentCurryVecEnv` also subclasses `CurryVecEnv`. It seems to be implemented very similarly to `CurryVecEnv`, so it's probably OK to merge them as well. Although if you can see a clean way of keeping them separate even better (I think this might be possible if we have `step_async` and `step_wait` call some instance helper methods, that can be overridden in subclasses).

Reply:
Okay. If you feel strongly that you prefer it, I'll work on writing the code for a big merged CurryVecEnv with transparency and multi-env embedding, even though it still strikes me as likely to make things harder to understand/maintain.

Reply:
I'm worried I'm just missing something here. Maybe you could explain a bit more why you think it's a bad idea? I'd much rather have a modular solution than everything in one big class. So if that's an option we should totally go for it. But the status quo does have one big merged class, `MultiCurryVecEnv`. If we can't make that simpler somehow, then it seems better to have just that than that + a subset of its functionality.

Reply:
One idea that might help with a more modular solution: we could make a wrapper for *policies* that takes a list of policies and randomizes which one makes a given prediction. This seems like it'd encompass most of the custom logic from `MultiCurryVecEnv`, and then we could just feed this multi-policy directly into `CurryVecEnv` and scrap `MultiCurryVecEnv`.

Reply:
I think it's a bad idea because if someone was trying to understand what the imagined MultiCompatibleCurryVecEnv is doing (name is obviously made up), in the majority of cases where they're just currying a single victim, they'd have to wade through a bunch of code for randomly assigning policies to different environments within a vecenv and keeping track of them and switching them at particular episode-break intervals, even though none of that will actually be being used in the case where we're using the same policy for all environments in a batched way. Basically, I think it's helpful for understanding the code to keep the simpler base case where we're just doing environment embedding of a single policy, rather than having the only way to do the simpler thing be to use the most flexibly-complex machinery.  

I think a problem with doing a policy-wrapper solution is the issue we discussed earlier around state: in cases where we have stateful policies (which it seems liked we'd want to make this work for, even though we haven't run stateful tests yet) we'd probably want to use a single policy for the duration of an episode so it can have a consistent on-distribution state, so you'd have to have some way of signaling to the policy-wrapper to change itself on episode boundaries, which does feel like it would make more sense to do on the level Env wrapper.